client/ayon_core/tools/utils/widgets.py

```diff
@@ -104,6 +105,151 @@ def __init__(self, *args, **kwargs):
+
+    def __init__(
+        self,
+        options: Optional[List[str]] = None,
+        parent: Optional[QtWidgets.QWidget] = None
+    ):
+        super().__init__(parent)
+
+        text_input = HintedLineEditInput(self)
+        options_button = HintedLineEditButton(self)
+        options_button.setIcon(get_down_arrow_icon())
+
+        main_layout = QtWidgets.QHBoxLayout(self)
+        main_layout.setContentsMargins(0, 0, 0, 0)
+        main_layout.setSpacing(0)
+        main_layout.addWidget(text_input, 1)
+        main_layout.addWidget(options_button, 0)
+
+        # Expand line edit and button vertically so they have same height
+        for widget in (text_input, options_button):
+            w_size_policy = widget.sizePolicy()
+            w_size_policy.setVerticalPolicy(
+                QtWidgets.QSizePolicy.MinimumExpanding)
+            widget.setSizePolicy(w_size_policy)
+
+        # Set size hint of this frame to fixed so size hint height is
+        #   used as fixed height
+        size_policy = self.sizePolicy()
+        size_policy.setVerticalPolicy(QtWidgets.QSizePolicy.Fixed)
+        self.setSizePolicy(size_policy)
+
+        text_input.returnPressed.connect(self.returnPressed)
+        text_input.textChanged.connect(self.textChanged)
+        text_input.textEdited.connect(self.textEdited)
+        options_button.clicked.connect(self._on_options_button_clicked)
+
+        self._text_input = text_input
+        self._options_button = options_button
+        self._options = None
+
+        # Set default state
+        self.set_options(options)
+
+    def text(self) -> str:
+        return self._text_input.text()
+
+    def setText(self, text: str):
+        self._text_input.setText(text)
+
+    def setPlaceholderText(self, text: str):
+        self._text_input.setPlaceholderText(text)
+
+    def placeholderText(self) -> str:
+        return self._text_input.placeholderText()
+
+    def setReadOnly(self, state: bool):
+        self._text_input.setReadOnly(state)
+
+    def setIcon(self, icon: QtGui.QIcon):
+        self._options_button.setIcon(icon)
+
+    def setToolTip(self, text: str):
+        self._text_input.setToolTip(text)
+
+    def set_button_tool_tip(self, text: str):
+        self._options_button.setToolTip(text)
+
+    def set_options(self, options: Optional[List[str]] = None):
+        self._options = options
+        self._options_button.setEnabled(bool(options))
+
+    def sizeHint(self) -> QtCore.QSize:
+        hint = super().sizeHint()
+        tsz = self._text_input.sizeHint()
+        bsz = self._options_button.sizeHint()
+        hint.setHeight(max(tsz.height(), bsz.height()))
+        return hint
+
+    # Adds ability to change style of the widgets
+    # - because style change of the 'HintedLineEdit' may not propagate
+    #   correctly 'HintedLineEditInput' and 'HintedLineEditButton'
+    def set_text_widget_object_name(self, name: str):
+        self._text_input.setObjectName(name)
```

Comment:
Can't we use Object Names for this in general? So that the inner object name is **always** "button" or "text" or whatever and when referencing them in the `qss` stylesheet we point to something like `HintedLineEdit button` or whatever? CSS allows such 'child object' referencing to dedicate styling for them? Or even without explicit object names we could even do something like `HintedLineEdit QPushButton:hover` etc. right?

Reply:
You would have to re-define all styles again. Even if you want to change just one thing.